src/vpn-worker.ts

```diff
@@ -105,32 +105,45 @@ const worker = async (instanceId: number, serviceId: number) => {
 	});
 
 	const drainConnections = async () => {
-		const clientCount = Object.keys(clientCache).length;
+		let clientCount = Object.keys(clientCache).length;
 		logger.info(`${clientCount} clients connected`);
 
 		if (clientCount > 0) {
-			logger.debug(JSON.stringify(clientCache));
 			const delayMs = DEFAULT_SIGTERM_TIMEOUT / clientCount;
 			logger.info(
 				`disconnecting ${clientCount} clients, spaced by ${delayMs}ms`,
 			);
 
 			let timeToKill = DEFAULT_SIGTERM_TIMEOUT;
 
-			for (let clientId = 0; clientId < clientCount; clientId++) {
+			for (const [key] of Object.entries(clientCache)) {
 				try {
-					logger.debug(JSON.stringify(clientCache[clientId]));
-					const cn = clientCache[clientId].uuid;
-					logger.info(`disconnecting ${cn}`);
-					// update device state
-					clients.setConnected(cn, serviceId, false, logger);
-					// disconnect client from VPN
-					await vpn.killClient(cn);
+					const _key = parseInt(key, 10);
+					try {
+						const cn = clientCache[_key].uuid;
+						logger.info(`disconnecting ${cn}`);
+
+						// disconnect client from VPN
+						try {
+							await vpn.killClient(cn);
+						} catch (err) {
+							logger.warning(
+								`${err} while killing ${cn} on worker ${serviceId}`,
+							);
+						}
+
+						// update device state
+						clients.setConnected(cn, serviceId, false, logger);
+					} catch {
+						// https://stackoverflow.com/questions/3633362/is-there-any-way-to-use-a-numeric-type-as-an-object-key/3633390#3633390
+						// https://bobbyhadz.com/blog/typescript-element-implicitly-any-type-not-number
+						// https://www.typescriptlang.org/tsconfig#suppressImplicitAnyIndexErrors
+					}
 				} catch (err) {
 					logger.warning(`received '${err}' trying to disconnect client`);
 				}
 				// last client disconnected
-				if (clientCount - clientId === 1) {
+				if (clientCount <= 1) {
```

Comment:
You should be able to just move this outside of the `for` loop and ignore client counting since you'll have gone through and attempted disconnections of them all at that point, ie they'll all have had a disconnect triggered